packages/front-end/components/Experiment/ResultsDownloadButton.tsx

```diff
@@ -0,0 +1,138 @@
+import {
+  ExperimentReportResultDimension,
+  ExperimentReportVariation,
+} from "back-end/types/report";
+import React, { useMemo } from "react";
+import { FaFileExport } from "react-icons/fa";
+import { useDefinitions } from "../../services/DefinitionsContext";
+import { getRisk } from "../../services/experiments";
+import { Parser } from "json2csv";
+import { MetricInterface, MetricStats } from "back-end/types/metric";
+
+type CsvRow = {
+  date?: string;
+  dimension?: string;
+  metricName?: string;
+  variantName?: string;
+  riskOfChoosing?: number;
+  users?: number;
+  count?: number;
+  conversionRate?: number;
+  chanceToBeatControl?: number | null;
+  percentChange?: number | null;
+};
+
+type Variation = {
+  buckets?: { x: number; y: number }[];
+  ci?: [number, number];
+  cr: number;
+  risk?: [number, number];
+  stats?: MetricStats;
+  users: number;
+  value: number;
+};
+
+type Row = {
+  label?: string;
+  metric?: MetricInterface;
+  rowClass?: string;
+  variations?: Variation[];
+};
+
+export default function ResultsDownloadButton({
+  results,
+  metrics,
+  variations,
+  trackingKey,
+  dimension,
+}: {
+  results?: ExperimentReportResultDimension[];
+  metrics: string[];
+  variations: ExperimentReportVariation[];
+  trackingKey: string;
+  dimension: string;
+}) {
+  const { getMetricById, getDimensionById, ready } = useDefinitions();
+
+  const json2csvParser = new Parser();
+
+  const csvRows: CsvRow[] = [];
+
+  const dimensionName = getDimensionById(dimension);
+
+  const getRows = () => {
+    const rows = [];
+    if (!results || !variations || !ready) return [];
+
+    results.forEach((result) => {
+      metrics.forEach((m) => {
+        const row: Row = {};
+        const metric = getMetricById(m);
+        row.label = metric?.name;
+        row.metric = metric;
+        row.rowClass = metric?.inverse ? "inverse" : "";
+        row.variations = result.variations.map((v) => {
+          return v.metrics[m];
+        });
+        rows.push(row);
+      });
+    });
+
+    rows.forEach((row, rowIndex) => {
+      row.variations.forEach((variation, index) => {
+        const updatedRow: CsvRow = {};
+        if (dimension === "pre:date") {
+          updatedRow.date = results[rowIndex].name;
+        } else if (results[rowIndex].name !== "All") {
```

Comment:
This won't work.  Up above, you are looking through results, and then looping through metrics in order to build the rows array.  So the length of rows is `results.length * metrics.length`.  You can;t use rowIndex like this to map to a result.

Reply:
Ahhh - really good catch. In my testing I've just been working with a single metric, so results.length hasn't been affected by metrics.length.

I lifted this logic up to the initial loop through results (and then metrics). It seems to be working with the limited data I've got in my env, but I'm going to continue poking at it to see if I can find any edge cases.